File: Introibo/Data/Hour.swift
import Foundation

// Matches Introibo/Resources/hours.json — the 8 canonical hours of the
// 1962 Roman Breviary.

struct Hour: Identifiable, Decodable, Hashable, ScheduledHour {
    let slug: String
    let name: String        // Latin name (Matutínum, Laudes, ...)
    // Vernacular metadata is `var` so the Spanish overlay can rewrite it
    // (ContentStore.applyVernacularOverlay). The parts stay Latin/English.
    var eng: String         // Vernacular name
    var time: String        // "at midnight", "at dawn", ...
    let hour: Int           // 0-23
    let minute: Int
    let glyph: String       // Single-letter dial glyph (M, L, I, III...)
    let order: Int          // Roman order for Hora I/II/...
    var intro: String       // Short prose introduction
    var parts: [Part]
    var related: [RelatedLink]? = nil

    var id: String { slug }

    // Heterogeneous parts. We decode into a sum-type-ish struct that
    // carries whichever fields are present; views switch on `type`.
    struct Part: Decodable, Hashable {
        let type: String
        var label: String? = nil
        var title: String? = nil
        var ref: String? = nil
        var lat: String? = nil
        var eng: String? = nil
        var latR: String? = nil
        var engR: String? = nil
        var v1Lat: String? = nil
        var v1Eng: String? = nil
        var r1Lat: String? = nil
        var r1Eng: String? = nil
        var v2Lat: String? = nil
        var v2Eng: String? = nil
        var r2Lat: String? = nil
        var r2Eng: String? = nil
        var verses: [Verse]? = nil
        var season: String? = nil
        var engBody: String? = nil
        var variationKey: String? = nil
        var antiphonLat: String? = nil
        var antiphonEng: String? = nil

        struct Verse: Decodable, Hashable {
            let lat: String
            var eng: String
        }
    }
}

// Matches Introibo/Resources/marian_antiphons.json.
struct MarianAntiphonData: Identifiable, Decodable, Hashable {
    let slug: String
    let title: String
    var eng: String
    let season: String
    let lat: String
    var engBody: String
    // The ℣/℟ and Orémus that follow the antiphon at Compline.
    var versicleLat: String?
    var versicleEng: String?
    var collectLat: String?
    var collectEng: String?
    // Alma Redemptóris only: the forms used from Christmas to Feb 1.
    var versicleLatChristmas: String?
    var versicleEngChristmas: String?
    var collectLatChristmas: String?
    var collectEngChristmas: String?

    var id: String { slug }

    enum CodingKeys: String, CodingKey {
        case slug, title, eng, season, lat, engBody
        case versicleLat = "versicle_lat"
        case versicleEng = "versicle_eng"
        case collectLat = "collect_lat"
        case collectEng = "collect_eng"
        case versicleLatChristmas = "versicle_lat_christmas"
        case versicleEngChristmas = "versicle_eng_christmas"
        case collectLatChristmas = "collect_lat_christmas"
        case collectEngChristmas = "collect_eng_christmas"
    }
}

File: android/app/src/main/java/com/lampstandhq/introibo/data/model/Hour.kt
package com.lampstandhq.introibo.data.model

import com.lampstandhq.introibo.data.links.RelatedLink
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Matches hours.json — the 8 canonical hours of the 1962 Roman Breviary.
 */
@Serializable
data class Hour(
    val slug: String,
    val name: String,        // Latin name (Matutinum, Laudes, ...)
    val eng: String,         // English name
    val time: String,        // "at midnight", "at dawn", ...
    val hour: Int,           // 0-23
    val minute: Int,
    val glyph: String,       // Single-letter dial glyph (M, L, I, III...)
    val order: Int,          // Roman order for Hora I/II/...
    val intro: String,       // Short prose introduction
    val parts: List<Part>,
    val related: List<RelatedLink>? = null,
) {
    /**
     * Heterogeneous parts. We decode into a sum-type-ish class that
     * carries whichever fields are present; views switch on [type].
     */
    @Serializable
    data class Part(
        val type: String,
        val label: String? = null,
        val title: String? = null,
        val ref: String? = null,
        val lat: String? = null,
        val eng: String? = null,
        val latR: String? = null,
        val engR: String? = null,
        val v1Lat: String? = null,
        val v1Eng: String? = null,
        val r1Lat: String? = null,
        val r1Eng: String? = null,
        val v2Lat: String? = null,
        val v2Eng: String? = null,
        val r2Lat: String? = null,
        val r2Eng: String? = null,
        val verses: List<Verse>? = null,
        val season: String? = null,
        val engBody: String? = null,
        val variationKey: String? = null,
        val antiphonLat: String? = null,
        val antiphonEng: String? = null,
    ) {
        @Serializable
        data class Verse(
            val lat: String,
            val eng: String,
        )
    }
}

/**
 * Matches marian_antiphons.json.
 */
@Serializable
data class MarianAntiphonData(
    val slug: String,
    val title: String,
    val eng: String,
    val season: String,
    val lat: String,
    val engBody: String,
    // The ℣/℟ and Orémus that follow the antiphon at Compline.
    @SerialName("versicle_lat") val versicleLat: String? = null,
    @SerialName("versicle_eng") val versicleEng: String? = null,
    @SerialName("collect_lat") val collectLat: String? = null,
    @SerialName("collect_eng") val collectEng: String? = null,
    // Alma Redemptóris only: the forms used from Christmas to Feb 1.
    @SerialName("versicle_lat_christmas") val versicleLatChristmas: String? = null,
    @SerialName("versicle_eng_christmas") val versicleEngChristmas: String? = null,
    @SerialName("collect_lat_christmas") val collectLatChristmas: String? = null,
    @SerialName("collect_eng_christmas") val collectEngChristmas: String? = null,
)
